torch/nn/modules/container.py

```diff
@@ -187,6 +187,104 @@ def extend(self, modules):
         return self
 
 
+class ModuleDict(Module):
+    r"""Holds submodules in a dictionary.
+
+    ModuleDict can be indexed like a regular Python dictionary, but modules it
+    contains are properly registered, and will be visible by all Module methods.
+
+    Arguments:
+        mapping (iterable, optional): an iterable of key/:class:`~torch.nn.Module``
+            pairs to add.
+
+    Example::
+
+        class MyModule(nn.Module):
+            def __init__(self):
+                super(MyModule, self).__init__()
+                self.choices = nn.ModuleDict({
+                        'conv': nn.Conv2d(10, 10, 3),
+                        'pool': nn.MaxPool2d(3)
+                })
+
+            def forward(self, x, choice):
+                x = self.choices[choice](x)
+                return x
+    """
+
+    def __init__(self, mapping=None):
+        super(ModuleDict, self).__init__()
+        if mapping is not None:
+            self.update(mapping)
+
+    def __getitem__(self, key):
+        return self._modules[str(key)]
+
+    def __setitem__(self, key, module):
+        self.add_module(str(key), module)
+
+    def __delitem__(self, key):
+        del self._modules[str(key)]
```

Comment:
We definitely shouldn't do `str` for someone. We can assert it's a string instead. It's ok for builtin types, but will become very confusing once someone starts using tensors or more complex objects, with larger `str` output.